fix(astro): resolve Astro's private APIs properly

AriPerkkio opened the pull request:
- Fixes https://github.com/stackblitz/tutorialkit/issues/235

The usage of private APIs was added in https://github.com/stackblitz/tutorialkit/pull/4 (👋 @d3lm).

Nemikolh reviewed (commented):
This API was added after we talked to them 😄 

Having that import is intentional here. Once it's stabilized on the Astro side, we'll be able to get rid of the type error.

AriPerkkio commented:
> we'll be able to get rid of the type error

This is not just type error. In monorepo setups this import actually fails and TutorialKit fails to start. Here's example from https://github.com/AriPerkkio/tutorialkit/tree/test/ui-tests/test/ui which doesn't work without this fix. 

<img src="https://github.com/user-attachments/assets/31085939-9de6-4d23-88a9-88b95e75187c" width="400" />

d3lm commented:
Yes, I agree with @Nemikolh and I'd like to keep it as is. The whole point to get this in upstream is to avoid copying code from Astro so we can fully rely on upstream for this. I think we can give them the feedback that it works well for us so it can be exposed properly and we don't need to import from `node_modules` directly. But I am not a fan of copying it again cause that's what we wanted to avoid in the first place.

AriPerkkio commented:
Sure, that would be ideal case. It's also breaking change as we would need to set new minimum version of `astro` but we can do breaking change release at this point easily.

If you have any ideas how to fix #235 and https://github.com/stackblitz/tutorialkit/pull/241 without this change, feel free to help! Otherwise this will block those two, I think.

Nemikolh commented:
@AriPerkkio For #235, I think we can temporarily use a config option. We should definitely provide feedback that this has been working fine for us.

For #241, why is this an issue there? Doesn't the dependency also get installed by `pnpm`? (I only had a quick look so sorry if it's something obvious 🙈)

AriPerkkio commented:
Removed copied Astro's `swap-functions.js` and added Vite plugin that resolves it properly. I tried to add module resolving logic directly to `Layout.astro` but it just didn't work. It's best to do on Vite's side.

Tested with https://github.com/stackblitz/tutorialkit/issues/235 and https://github.com/stackblitz/tutorialkit/pull/241. Checked dev mode and build `dist` contents. 

Using private APIs like this relies on Astro's npm package's file structure and can break at any time. There's reason why `package.json`'s `exports` are required for imports. In my opinnion it would be best to simply copy-paste this from Astro if it's wanted to be used. And once they include it in their public API (`exports`), we can start requiring that version as peer dep.

d3lm commented:
> We should definitely provide feedback that this has been working fine for us.

@Nemikolh @AriPerkkio I can provide that feedback to them and ask them to expose it properly for us.

Nemikolh commented on packages/astro/src/vite-plugins/astro-swap-functions.ts:
```diff
@@ -0,0 +1,36 @@
+import type { VitePlugin } from '../types.js';
+
+const require = createRequire(import.meta.url);
+const astroDist = resolve(require.resolve('astro/package.json'), '..');
```
Oh interesting, why do we `require.resolve` `astro/package.json` instead of just `astro`? 👀

AriPerkkio replied:
Resolving `astro` gives us the entrypoint of the package. Here we want to instead look exactly for `dist/transitions` directory.

```
> require.resolve("astro")
'/Users/x/tutorialkit/node_modules/.pnpm/astro@4.12.2_@types+node@20.14.11_typescript@5.5.3/node_modules/astro/dist/core/index.js'
> require.resolve("astro/package.json")
'/Users/x/tutorialkit/node_modules/.pnpm/astro@4.12.2_@types+node@20.14.11_typescript@5.5.3/node_modules/astro/package.json'
```

We could of course assume that their entrypoint always stays at `astro/dist/core/index.js`, and traverse the path up from there. 🤔

Nemikolh replied:
Ah thanks for the explanation!

Nemikolh commented on packages/astro/src/vite-plugins/astro-swap-functions.ts:
```diff
@@ -0,0 +1,36 @@
+
+    async load(id) {
+      if (id === resolvedId) {
+        return await readFile(swapFunctionEntry, 'utf8');
```
Wouldn't it be better to return an import to that path? If the module import other files, those import will fail if we return its content like this.

AriPerkkio replied:
I haven't heard of this problem before. How would that look in practice here? 

Is this error mentioned somewhere here https://rollupjs.org/plugin-development/#load?

Nemikolh replied:
Ah no worries! I'll clarify. 

Consider relative imports. There is no way they can work as the module path is lost when just returning the module content as a virtual module. If you look at what we did for the `theme.css` file, we import it (instead of including its content) precisely to support local imports.

AriPerkkio replied:
Oh good point, I didn't realize this! 💯 

I'll change this to `export * from '${swapFunctionEntry}';`. There's no `default` export so maybe we don't need to handle that one here.

Nemikolh commented on packages/astro/src/default/layouts/Layout.astro:
```diff
@@ -38,7 +38,8 @@ const baseURL = import.meta.env.BASE_URL;
     </script>
     <script>
-      import * as builtInSwap from 'node_modules/astro/dist/transitions/swap-functions';
+      // @ts-ignore -- untyped
```
We can add an entry for that module in `env-default.d.ts` to avoid the `//@ts-ignore`.

(we used to have a type error here though that we ignored so maybe this is fine)

AriPerkkio replied:
I think it's better to keep `@ts-ignore`. By defining types in declaration file, we would need to somehow try to keep them up-to-date with Astro's implementation.

Nemikolh replied:
Oh I was thinking of declaring the module as exporting `any` like we do for CSS modules.

AriPerkkio replied:
But that would require specifying each named export in the declaration, right? There's no way to say dynamically "all named exports from here are any".

Nemikolh replied:
You don't have to. You can do like what we do for CSS modules:

```ts
declare module 'tutorialkit:astro-swap-functions';
```

AriPerkkio replied:
Nice - happy to learn new tricks like this! 💯

Nemikolh reviewed (commented):
Looking good! 🤩

d3lm commented on packages/astro/src/default/layouts/Layout.astro:
```diff
@@ -38,7 +38,8 @@ const baseURL = import.meta.env.BASE_URL;
     </script>
     <script>
-      import * as builtInSwap from 'node_modules/astro/dist/transitions/swap-functions';
+      // @ts-ignore -- untyped
```
```suggestion
      // @ts-ignore because it's untyped
```

d3lm commented on packages/astro/src/vite-plugins/astro-swap-functions.ts:
Clever!

d3lm commented:
Ok PR is on the way https://github.com/withastro/astro/pull/11708

Nemikolh commented:
@d3lm, @AriPerkkio do we want to wait for that PR to land on Astro's side then? Or are we all happy to move forward with this PR?

AriPerkkio commented:
Let's close this and wait for Astro's PR instead. I can add work-around into the #241 for now, and post same instructions in #235.
